New Cropper across the app

ItsEeleeya opened the pull request:
This RP removes the two previous cropper implementation and replaces them with a new, feature rich, more efficient unified cropper component.

![Unified Cropper](https://github.com/user-attachments/assets/86d8a1b2-3729-4803-b844-3740203fee47)


The new cropper no longer uses a canvas and consumes much less resources in comparison.

Changes:
- Removed CropAreaRenderer.
- Updated the look and animation of the capture-area controls, inspired by the header in Cap's main webpage.
- Removed the option for disabling haptics as it can be changed from the System Settings.
- Use the correct NSPopUpMenuWindowLevel as the layer for the capture-area window.
- Add the default capability `core:resources:allow-close` for properly closing context menus.
- The ErrorBoundary view now has a button to try and close the window as it might not always be possible without quitting the app.
- The new cropper retains the previous features except for drag/resize using the touchpad or control key and mouse scroll.
- Double clicking any handle will try to fill the area available on that side.
- The cropper now smoothly animates changes done via buttons.
- The crop selection can be flipped.
- A new selection can be made by clicking and dragging across an area outside of the selection.
- The selection size is now shown near the cursor during resize (Except on the editor).
- Editor crop dialog now has a fill button.
- Double clicking any area in the cropper now expands the selection as much as is possible while respecting constraints.
- Menus in target-select-overlay now open with a set position to appear directly below their respective button.

With this, I tried having it resemble most of the functionality that is available in the cropper that's part of macOS's native Screenshot app.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Major cropping overhaul: richer crop UI with aspect/snap menu, per-display persisted crops, reset/fill/confirm controls, and context-menu crop options.

* **Bug Fixes**
  * Setup and capture-area windows are now non-resizable; improved macOS stacking for capture area.
  * Opening the crop dialog now stops playback before showing it.
  * Error UI: added a destructive "Close" button to dismiss the window on fatal errors.

* **Chores**
  * Removed legacy haptics toggle and legacy canvas/box/crop-overlay utilities; added permission to allow closing the window.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ItsEeleeya commented on apps/desktop/src/components/CapErrorBoundary.tsx:
```diff
@@ -32,6 +33,15 @@ export function CapErrorBoundary(props: ParentProps) {
 								Reload
 							</Button>
+							<Button
+								onClick={() => {
+									getCurrentWebviewWindow().close();
+									window.close();
+								}}
+								variant="destructive"
+							>
+								Try Close
+							</Button>
```
The function is sync:
```
/**
 * Get an instance of `Webview` for the current webview window.
 *
 * @since 2.0.0
 */
declare function getCurrentWebviewWindow(): WebviewWindow;
```

ItsEeleeya commented on apps/desktop/src/routes/capture-area.tsx:
```diff
@@ -1,78 +1,110 @@
+
+	const [crop, setCrop] = createSignal(CROP_ZERO);
+
+	const scheduled = createScheduled((fn) => debounce(fn, 50));
+
+	const isValid = createMemo((p: boolean = true) => {
+		const b = crop();
+		return scheduled()
+			? b.width >= MIN_SIZE.width && b.height >= MIN_SIZE.height
+			: p;
 	});
```
The function returns `Accessor<boolean>`:
```
export declare function createScheduled(schedule: (callback: VoidFunction) => VoidFunction): Accessor<boolean>;
```
Usage is correct as shown in the [solid-primitives/scheduled](https://github.com/solidjs-community/solid-primitives/tree/main/packages/scheduled#createscheduled) docs.

ItsEeleeya commented on apps/desktop/src/routes/editor/ConfigSidebar.tsx:
```diff
@@ -1423,14 +1421,14 @@ function BackgroundConfig(props: { scrollRef: HTMLDivElement }) {
 																	if (
 																		!moveEvent.shiftKey &&
-																		hapticsEnabled() &&
+																		hapticsEnabled &&
 																		project.background.source.type ===
 																			"gradient" &&
 																		project.background.source.angle !== newAngle
 																	) {
 																		commands.performHapticFeedback(
-																			"Alignment",
-																			"Now",
+																			"alignment",
+																			"now",
 																		);
 																	}
```
It is not reactive and it's not a function. The function `type()` from `@tauri-apps/plugin-os` is also sync.

ItsEeleeya commented on apps/desktop/src/routes/target-select-overlay.tsx:
```diff
@@ -1,26 +1,34 @@
+import { createResizeObserver } from "@solid-primitives/resize-observer";
 import { useSearchParams } from "@solidjs/router";
 import { createQuery } from "@tanstack/solid-query";
+import { LogicalPosition } from "@tauri-apps/api/dpi";
```
That's incorrect, the one in ~/utils/tauri is simply a type. LogicalPosition is a class that we have to use with the menu function from @tauri-apps/api/dpi.

ItsEeleeya commented on apps/desktop/src/routes/capture-area.tsx:
```diff
@@ -90,55 +121,131 @@ export default function CaptureArea() {
-						<div class="transition-all ease-out duration-200 absolute w-auto h-10 bg-gray-1 rounded-[12px] drop-shadow-2xl overflow-visible border border-gray-3 outline outline-1 outline-gray-6 flex justify-around p-1 top-11">
+						<div
+							class="scale-100 flex items-center p-1 gap-2 absolute w-auto h-14 rounded-full top-12"
+							classList={{ "flex-row-reverse": ostype() === "windows" }}
+						>
```
That's not correct, the function returns OsType which is:
```
type OsType = 'linux' | 'windows' | 'macos' | 'ios' | 'android';
```

ItsEeleeya commented on apps/desktop/src/routes/capture-area.tsx:
```diff
@@ -147,16 +254,25 @@ export default function CaptureArea() {
 					<Cropper
-						class="transition-all duration-200"
-						value={crop}
+						ref={cropperRef}
+						aspectRatio={aspect() ?? undefined}
```
it works

ItsEeleeya replied:
it does work

ItsEeleeya commented on apps/desktop/src/routes/capture-area.tsx:
```diff
@@ -90,55 +121,131 @@ export default function CaptureArea() {
+										title="Aspect Ratio"
+										class="group flex items-center justify-center size-10 text-gray-11 hover:bg-gray-5 active:bg-gray-6 rounded-full transition-colors duration-200 cursor-default"
+										type="button"
+										onMouseDown={showCropOptionsMenu}
+										onClick={showCropOptionsMenu}
+									>
```
No it doesn't. Menus should open on mouse down. onClick is only for keyboard or touch.

oscartbeaumont commented on apps/desktop/src/utils/tauriSpectaHack.ts:
```diff
@@ -0,0 +1,24 @@
+	event: K,
+	target: string,
+	value: (typeof events)[K] extends __EventObj__<infer U> ? U : never,
+) => webview.emitTo(target, event, value);
```
Good catch CodeRabbit. Will fix this before merging!

ItsEeleeya commented:
This should be ready now!
I was going to do some more changes that'd alter the design and the UX for this, but I think it'll be better for something later 👀

> When a fixed ratio is set and you drag the mouse to one side of the screen the box will start resizing from the opposite side. It should just stop resizing as you keep dragging offscreen. [Demo](https://cap.so/s/yyd9qk3v3eva3tf)

This is now fixed!

> I managed to trigger a SolidJS error in Studio mode by applying a crop and pressing save. Demo

I still have no idea how it occurred and I couldn't reproduce it.

> When the selection area is smaller than the selected range we would rather have a warning than do the snapping behavior. This is what is currently implemented on main with my new changes so we should backport it to this PR.

It's been added back but adjust the look and wording as you see fit.

> Work through CodeRabbit suggestion and solve them where they are relevant
As much I wish it had something relevant, none of them are in my case.

I spent over a day trying to figure out why animations kept breaking for a new indicator I added, nothing helped until I randomly came across an evil `stopPropagation` on the keyup event. anyway I just decided to not add an animation because of issues I had with the Transition group, or maybe an skill issue.
It was going to be an animation similar to that of the control bar on capture-area, but classes weren't applying normal, transition mode would break after a while etc.


I was also looking at #1282 
and a couple of weeks ago I got this concept for it:
_Which is possible on macOS 26, and possibly even older on the webview, no native view_
![cap-bar](https://github.com/user-attachments/assets/7d6362eb-b340-40a7-87b2-7a737666378c)

Althouth @ameer2468 said to me that there isn't yet a plan to make a change like this.

I was thinking there could just be a floating menu (similar to the one on capture-area) to select the target mode.

Brendonovich reviewed (approved):
lgtm 👍
